### PR TITLE
modules: add env option

### PR DESCRIPTION
**Pull request by GaetanLepage**
Add a top-level `env` option that allows to propagate environment variables in the wrapped neovim package.

```nix
env = {
  FOO = "1";
  BAR_PATH = "/home/me/.local/share/bar";
  BAZ_MAX_COUNT = "1000";
}
```

Fixes #3055

### CODE REVIEW

**Review by traxys — APPROVED** [review]

File: modules/top-level/output.nix [hunk 1]
```diff
@@ -339,7 +339,9 @@ in
           initSource;
 
       extraWrapperArgs = builtins.concatStringsSep " " (
-        (optional (
+        # Setting environment variables in the wrapper
+        (lib.mapAttrsToList (var: value: "--set ${var} ${value}") config.env)
```
Comment (MattSturgeon) [inline modules/top-level/output.nix:343]: Do we need `lib.escapeShellArg` here?

File: tests/test-sources/env.nix [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
+  env-variables = {
+    env = {
+      FOO = "1";
+      BAR_PATH = "/home/me/.local/share/bar";
+      BAZ_MAX_COUNT = "1000";
+    };
+
```
Comment (MattSturgeon) [inline tests/test-sources/env.nix:8]: Let's add some tests for edge-cases, such as variable names or values that may need `lib.escapeShellArg`

File: modules/output.nix [hunk 3]
```diff
@@ -28,6 +28,17 @@ let
 in
 {
   options = {
+    env = mkOption {
+      type = with types; attrsOf str;
```
Comment (MattSturgeon) [inline modules/output.nix:32]: Down the line we could expand the type to also accept e.g. `path` and `package`, but this is fine for an initial starting point.

Idk what type the NixOS or home-manager `environment` option uses?

**Review by MattSturgeon — COMMENTED** [review]
Awesome 😎

File: modules/top-level/output.nix (hunk 1 above)
Comment (MattSturgeon) [reply modules/top-level/output.nix:343]: I'm wondering if `escapeShellArgs` would be neater:

```nix
lib.escapeShellArgs [ "--set" name value ]
```

File: modules/output.nix (hunk 3 above)
Comment (khaneliman) [reply modules/output.nix:32]: I thought environment variables could only be strings?

Comment (khaneliman) [reply modules/output.nix:32]: nvm https://github.com/nixos/nixpkgs/blob/master/nixos/modules/config/shells-environment.nix#L41-L41

Comment (MattSturgeon) [reply modules/output.nix:32]: I'm fine with keeping things simple in this PR and opening an issue regarding the other types we could support

**Review by khaneliman — APPROVED** [review]
LGTM

File: modules/top-level/output.nix [hunk 4]
```diff
@@ -339,7 +339,9 @@ in
           initSource;
 
       extraWrapperArgs = builtins.concatStringsSep " " (
-        (optional (
+        # Setting environment variables in the wrapper
+        (lib.mapAttrsToList (var: value: "--set ${var} ${lib.escapeShellArg value}") config.env)
```
Comment (khaneliman) [inline modules/top-level/output.nix:343]: or should this be `escapeShellArgs` for the entire list? not too familiar with its usage

Comment (MattSturgeon) [reply modules/top-level/output.nix:343]: I'd personally prefer this:

```suggestion
        (lib.mapAttrsToList (name: value: lib.escapeShellArgs [ "--set" name value ]) config.env)
```

**Review by MattSturgeon — APPROVED** [review]
:rocket:

File: modules/output.nix [hunk 5]
```diff
@@ -28,6 +28,26 @@ let
...
   options = {
+    env = mkOption {
+      type =
+        with types;
+        lazyAttrsOf (oneOf [
+          str
+          path
+          int
+          float
+        ]);
```
Comment (MattSturgeon) [inline modules/output.nix:39]: Just a note:

The [nixos impl](https://github.com/NixOS/nixpkgs/blob/3acdee779318c707cfb5735db379b735f314b094/nixos/modules/config/shells-environment.nix#L42-L44) has an `apply` function that coerces path values to `"${value}"` (ensuring they get copied to the nix store) and also does some stuff with list values (although we don't currently support list types).

```nix
apply =
  let
    toStr = v: if lib.isPath v then "${v}" else toString v;
  in
  lib.mapAttrs (n: v: if lib.isList v then lib.concatMapStringsSep ":" toStr v else toStr v);
```

**Comment by MattSturgeon** [comment]
@mergify backport nixos-24.11
